distribution/lib/Standard/Base/0.0.0-dev/src/Data/XML.enso

```diff
@@ -0,0 +1,389 @@
+        XML_Error.handle_java_exceptions <|
+            XML_Utils.outerXML self.java_element
+
+    ## Gets the raw XML of the contents of the element, not including the
+       outermost tag and attributes.
+
+       > Example
+         Get the inner XML of an element.
+
+             XML_Document.from_string '<foo><bar>hello</bar></foo>' . root_element . inner_xml
+             # => '<bar>hello</bar>'
+    inner_xml : Text ! XML_Error
+    inner_xml self =
+        XML_Error.handle_java_exceptions <|
+            XML_Utils.innerXML self.java_element
+
+    ## Gets elements matching a given tag name.
+
+       This searches through all descendants of the node, not just immediate children.
+
+       > Example
+             XML_Document.from_string '<foo> <baz>hello</baz> <bar>and</bar> <baz>goodbye</baz> </foo>' . root_element . get_elements_by_tag_name "baz"
+             # => [XML_Document.from_string "<baz>hello</baz>" . root_element, XML_Document.from_string "<baz>goodbye</baz>" . root_element]
+    get_elements_by_tag_name : Text -> Vector XML_Element ! XML_Error
+    get_elements_by_tag_name self tag_name:Text =
+        XML_Error.handle_java_exceptions <|
+            only_wanted_nodes (nodelist_to_vector (self.java_element.getElementsByTagName tag_name))
+
+    ## PRIVATE
+    Value (java_element:Element)
+
+type XML_Error
+    # An error that indicates that the XML data could not be parsed.
+
+      Arguments:
+      - line_number: the line on which the parse failed.
+      - column_number: the column at which the parse failed.
+    Parse_Error (line_number : Integer) (column_number : Integer)
+
+    # Any other XML-related Java exception.
+    Other (error : Text)
+
+    ## PRIVATE
+       Convert the XML_Error to a human-readable format.
+    to_display_text : Text
+    to_display_text self = case self of
+        XML_Error.Parse_Error line_number column_number -> "The XML document could not be parsed at line " + line_number.to_text + ", column " + column_number.to_text
+        XML_Error.Other error -> error
+
+    ## PRIVATE
+
+       Utility method for running an action with Java exceptions mapping.
+    handle_java_exceptions : Any -> Any ! XML_Error
+    handle_java_exceptions ~action =
+        Panic.catch JException action caught_panic->
+            XML_Error.wrap_java_exception caught_panic.payload
+
+    ## PRIVATE
+
+       Converts a Java `Exception` into its Enso counterpart.
+    wrap_java_exception : JException -> XML_Error
+    wrap_java_exception exception:JException = case exception of
+        _ : SAXParseException -> Error.throw (XML_Error.Parse_Error exception.getLineNumber exception.getColumnNumber)
+        _ -> Error.throw (XML_Error.Other "An Exception has occurred: "+exception.to_text)
+
+## PRIVATE
+   Filter out unwanted nodes.
+   Wanted nodes are:
+   - Elements
+   - Text (if not completely whitespace)
+only_wanted_nodes : Vector Any -> Vector (Text | XML_Element)
+only_wanted_nodes nodes:(Vector Any) =
```

Comment:
think nodelist is always called with this so I'd wrap the two together.

Reply:
Done